tools/slang-capability-generator/capability-generator-main.cpp

```diff
@@ -65,21 +111,186 @@ struct CapabilityDef : public RefObject
+                this->keyAtomsPresent.add(keyAtom);
+            for (auto keyAtom : sharedKeyAtomsInCanonicalSet_stage)
+                this->keyAtomsPresent.add(keyAtom);
+        }
+        if (auto base = this->getAbstractBase())
+            keyAtomsPresent.add(this);
+    }
+};
+
+template<typename T, Index N>
+class PreallocatedStackVector
+{
+    T m_data[N];
+    Index m_currentElementCount = 0;
+    const Index m_maxElementCount = N;
+
+public:
+    struct Iterator
+    {
+        PreallocatedStackVector<T, N>& m_context;
+        Index m_element;
+
+        Iterator(PreallocatedStackVector<T, N>& context, Index element) : m_context(context), m_element(element)
+        {
+        }
+
+        Iterator begin() const
+        {
+            return m_context->begin();
+        }
+        Iterator end() const
+        {
+            return m_context->end();
+        }
+        Iterator& operator++()
+        {
+            this->m_element++;
+            return *this;
+        }
+        Iterator& operator++(int)
+        {
+            return ++(*this);
+        }
+        Iterator& operator--()
+        {
+            this->m_element--;
+            return *this;
+        }
+        bool operator==(const Iterator& other) const
+        {
+            return &this->m_context == &other.m_context && this->m_element == other.m_element;
+        }
+        bool operator!=(const Iterator& other) const
+        {
+            return !(*this == other);
+        }
+
+        const T& operator*() const
+        {
+            return m_context[m_element];
+        }
+        
+        T& operator*()
+        {
+            return m_context[m_element];
+        }
+    };
+
+    PreallocatedStackVector()
+    {
+    }
+
+
+    T& operator[](Index index)
+    {
+        return m_data[index];
+    }
+
+    Iterator begin()
+    {
+        return Iterator(*this, 0);
+    }
+
+    Iterator end()
+    {
+        return Iterator(*this, m_currentElementCount - 1);
+    }
+
+    T& getLast()
+    {
+        return (*this)[m_currentElementCount - 1];
+    }
+
+    Index getCount()
+    {
+        return m_currentElementCount;
+    }
+
+    void add(const T& item)
+    {
+        SLANG_ASSERT(m_currentElementCount != m_maxElementCount);
+
+        (*this)[m_currentElementCount] = item;
+        m_currentElementCount++;
+    }
 };
 
 struct CapabilityDefParser
 {
-    CapabilityDefParser(Lexer* lexer, DiagnosticSink* sink)
+    CapabilityDefParser(
+        Lexer* lexer, 
+        DiagnosticSink* sink, 
+        PreallocatedStackVector<CapabilityDef, kCapabilityDefCount>& defs)
         : m_lexer(lexer)
         , m_sink(sink)
+        , m_defs(defs)
     {
     }
     
     Lexer* m_lexer;
     DiagnosticSink* m_sink;
 
     Dictionary<String, CapabilityDef*> m_mapNameToCapability;
-    List<RefPtr<CapabilityDef>> m_defs;
+    PreallocatedStackVector<CapabilityDef, kCapabilityDefCount>& m_defs;
 
```

Comment:
I don't quite understand why this optimization is necessary. This is just generator code that runs once during build. As long as we are not incredibly slow, performance doesn't matter and we should prefer simplicity over low level optimizations. List<RefPtr<>> is fine here.

Reply:
I do not have a preference here.
@jkwak-work, any opinion on the matter?

Reply:
You need to bring it back to use `new`.
When I asked if we can avoid using `new`, your answer was `because it is too big for the stack`.
So I told you that it is not as big as you think.

But now I was looking at the code around it, it has to be allocated from the heap.
It is not because it is too big, but because the object should be alive after this function call is ended.

I don't remember telling you to use stack.
I was only asking if we can avoid unnecessary use of heap if possible.
In this case, it is not possible.

Reply:
You need to be aware of the different characteristics of stack and heap.
I assume your experience is mostly with Java and Java doesn't have an explicit way to use stack.
In C++, the use is always preferred but it is not always possible.

Reply:
>You need to be aware of the different characteristics of stack and heap.
 
I am very aware of `stack` vs. `heap`,  I primarily use C++/C for development, I understand the differences.

I believed "Can it be a stack memory?" was a suggestion to use stack.
* when I said I thought it was too big I was referring to how it is dynamic in size, but this remark seemed to be disregarded?

Reply:
Seems like there was some misunderstanding here, but I think we are on the same page now that it is OK to just heap allocate `CapabilityDef` and store them in RefPtrs.

The `CapabiltiyDef` is really an AST node here, and it is fine and convenient to just use heap allocation.

Reply:
Yes, there was a miscommunication between me and Ariel. I had a very long conversation with Ariel about this. It was my fault that my question and instruction were not detailed enough and it lead Ariel to believe to make changes in a way that any experienced programmers wouldn't prefer.

I will try to make my questions and instructions more clear next time to avoid a miscommunications like this. For this specific case, I should have asked my question in a following, as an example.
```
Can we change the function local variable to be allocated in the stack memory in the function that the variable is created from. And keep its member variables as it is in heap. And don't make changes to the List variable that the object will be added into.
```
Maybe that's too long but basically my instruction was not clear.